### PR TITLE
V2 configuration

### PR DESCRIPTION
**Pull request by nid90**
**Closes:** https://github.com/tramlinehq/tramline/issues/277, https://github.com/tramlinehq/tramline/issues/276

### CODE REVIEW

File: app/models/bitbucket_integration.rb [hunk 1]
```diff
@@ -277,7 +277,9 @@ def enable_auto_merge!(_pr_number)
 
   def workflows
     return [] unless integration.ci_cd?
-    with_api_retries { installation.list_pipeline_selectors(code_repository_name) }
+    cache.fetch(workflows_cache_key, expires_in: 10.minutes) do
```
Comment (nid90) [inline app/models/bitbucket_integration.rb:280]: for later: button on UI to refresh this cache

File: app/models/bitrise_integration.rb [hunk 2]
```diff
@@ -104,7 +104,9 @@ def metadata
 
   def workflows
     return [] unless integration.ci_cd?
-    installation.list_workflows(project, WORKFLOWS_TRANSFORMATIONS)
+    cache.fetch(workflows_cache_key, expires_in: 10.minutes) do
```
Comment (nid90) [inline app/models/bitrise_integration.rb:107]: for later: button on UI to refresh this cache

File: app/models/config/release_platform.rb [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  belongs_to :release_platform
+  has_one :internal_workflow, -> { internal }, class_name: "Config::Workflow", inverse_of: :release_platform_config, dependent: :destroy
+  has_one :release_candidate_workflow, -> { release_candidate }, class_name: "Config::Workflow", inverse_of: :release_platform_config, dependent: :destroy
+  has_one :internal_release, -> { internal }, class_name: "Config::ReleaseStep", inverse_of: :release_platform_config, dependent: :destroy
+  has_one :beta_release, -> { beta }, class_name: "Config::ReleaseStep", inverse_of: :release_platform_config, dependent: :destroy
+  has_one :production_release, -> { production }, class_name: "Config::ReleaseStep", inverse_of: :release_platform_config, dependent: :destroy
+
+  accepts_nested_attributes_for :internal_workflow, allow_destroy: true
+  accepts_nested_attributes_for :release_candidate_workflow, allow_destroy: true
```
Comment (nid90) [inline app/models/config/release_platform.rb:21]: we should remove the `allow_destroy: true` for this one

File: app/models/config/release_platform.rb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  accepts_nested_attributes_for :internal_release, allow_destroy: true
+  accepts_nested_attributes_for :beta_release, allow_destroy: true
```
Comment (nid90) [inline app/models/config/release_platform.rb:23]: we should remove the `allow_destroy: true`

File: app/models/config/submission.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+class Config::Submission < ApplicationRecord
+  self.table_name = "submission_configs"
+  FULL_ROLLOUT_VALUE = BigDecimal("100")
+
+  belongs_to :release_step_config, class_name: "Config::ReleaseStep"
+  has_one :submission_external, class_name: "Config::SubmissionExternal", inverse_of: :submission_config, dependent: :destroy
+
+  accepts_nested_attributes_for :submission_external, allow_destroy: true
+
+  validates :submission_type, presence: true
```
Comment (nid90) [inline app/models/config/submission.rb:24]: should validate uniqueness for number also

File: app/models/config/submission.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      errors.add(:rollout_stages, :increasing_order)
+    end
+
+    if rollout_stages.any? { |value| value > FULL_ROLLOUT_VALUE }
+      errors.add(:rollout_stages, :max_100)
+    end
+  end
+
+  def rolloutable?
+    rollout_enabled? && submission_type.in?(%w[PlayStoreSubmission])
```
Comment (nid90) [inline app/models/config/submission.rb:79]: AppStoreSubmussion is also rolloutable, no? not configurable, but rolloutable. we should be storing the default sequence in the db for it.

File: app/models/github_integration.rb [hunk 7]
```diff
@@ -231,7 +232,9 @@ def pr_open?(pr)
 
   def workflows
     return [] unless integration.ci_cd?
-    installation.list_workflows(code_repository_name, WORKFLOWS_TRANSFORMATIONS)
+    cache.fetch(workflows_cache_key, expires_in: 10.minutes) do
```
Comment (nid90) [inline app/models/github_integration.rb:235]: for later: button on UI to refresh this cache

File: app/models/config/submission.rb (hunk 6 above)
Comment (nid90) [reply app/models/config/submission.rb:79]: in other words, we should validate the rollout stages for all submission types if rollout is enabled

File: app/controllers/config/release_platforms_controller.rb [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+  def update_production_release_rollout_stages(production_release_attributes)
+    submission_attributes = production_release_attributes[:submissions_attributes]["0"]
+    if production_release_attributes.present? && submission_attributes[:rollout_enabled] == "true"
+      submission_attributes[:rollout_stages] = submission_attributes[:rollout_stages].safe_csv_parse
+    end
+  end
+
+  def find_workflow_name(identifier)
+    @ci_actions.find { |action| action[:id] == identifier }&.dig(:name)
```
[the remark this answers is not in the document]
Comment (kitallis) [reply app/controllers/config/release_platforms_controller.rb:144]: when would this be?

File: app/controllers/config/release_platforms_controller.rb [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  end
+
+  def find_submission_name(submission)
+    identifier = submission.dig(:submission_external_attributes, :identifier)
+    return unless identifier
+
+    @submission_types.find { |type| type[:type].to_s == submission[:submission_type].to_s }
+      &.then { |sub| sub.dig(:channels) }
+      &.then { |channels| channels.find { |channel| channel[:id].to_s == identifier } }
+      &.then { |channel| channel[:name] }
```
[the remark this answers is not in the document]
Comment (kitallis) [reply app/controllers/config/release_platforms_controller.rb:154]: when would this be?

File: app/controllers/config/release_platforms_controller.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (kitallis) [reply app/controllers/config/release_platforms_controller.rb:144]: not doing this right now, it's too tricky and rare

File: app/controllers/config/release_platforms_controller.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (kitallis) [reply app/controllers/config/release_platforms_controller.rb:154]: not doing this right now, it's too tricky and rare

**Pull request merged**